### PR TITLE
Add new mlxconfig params to mellanox_fw_update.yaml

### PR DESCRIPTION
**Pull request by wmousa**
(ESWITCH_HAIRPIN_TOT_BUFFER_SIZE, ESWITCH_HAIRPIN_DESCRIPTORS)

### CODE REVIEW

**Comment by wmousa** [comment]
@adrianchiris @moshe010 please review this

File: templates/firstboot/mellanox_fw_update.yaml [hunk 1]
```diff
@@ -40,6 +40,21 @@ parameters:
...
     default: False
     description: 'Priority tag required'
+  ESWITCH_HAIRPIN_TOT_BUFFER_SIZE:
+    type: json
+    default: {"*": "17"}
+    description: If "*" is on of the json keys, then its value will set to all indexes
+                 If you need to set configuration for specific indexes you can pass the
+                 value as below for index 2 to be 17 and index 3 to be 16
+                 Example, {"2": "17", "3": "16"}
+                 Make sure to choose only the PCI ends with 0
```
Comment (adrianchiris) [inline templates/firstboot/mellanox_fw_update.yaml]: >  Make sure to choose only the PCI ends with 0

if its a global/per-host parameter it shouldnt matter right ?
also if you run on more than one PF, we are not appying the configuration twice correct?

File: templates/firstboot/mellanox_fw_update.yaml [hunk 2]
```diff
@@ -40,6 +40,21 @@ parameters:
     type: boolean
     default: False
     description: 'Priority tag required'
+  ESWITCH_HAIRPIN_TOT_BUFFER_SIZE:
+    type: json
+    default: {"*": "17"}
```
Comment (adrianchiris) [inline templates/firstboot/mellanox_fw_update.yaml:45]: better than a list :)

File: templates/firstboot/mellanox_fw_update.yaml [hunk 3]
```diff
@@ -230,7 +263,30 @@ resources:
                                       " please update to the latest mstflint package." % key)
                             continue
 
-                        if current_mlx_config.get(key) and value.lower(
+                        if key in self.arrayed_mlnx_config.keys():
```
Comment (moshe010) [inline templates/firstboot/mellanox_fw_update.yaml]: remove the keys()
if key in self.arrayed_mlnx_config

File: templates/firstboot/mlnx_fw_update.py [hunk 4]
```diff
@@ -16,25 +16,28 @@
 FW_VERSION_REGEX = r'FW Version:\s*\t*(?P<fw_ver>\d+\.\d+\.\d+)'
 RUNNING_FW_VERSION_REGEX = r'FW Version\(Running\):\s*\t*(?P<fw_ver>\d+\.\d+\.\d+)'
 PSID_REGEX = r'PSID:\s*\t*(?P<psid>\w+)'
+ARRAY_VALUE_REGEX = r'Array\[(?P<first_index>\d+)\.\.(?P<last_index>\d+)\]'
```
Comment (moshe010) [inline templates/firstboot/mlnx_fw_update.py:20]: not for this PR. but in the future we should create the Yaml file automatically from this python script using jinja to render from template. Then you should maintain the script in the yaml file. Again this is not for now, but for future enchantment

Comment (adrianchiris) [reply templates/firstboot/mlnx_fw_update.py:20]: tbh the script is getting pretty large. if it keeps growing we would need to explore alternatives to just putting it flat in the yaml

File: templates/firstboot/mellanox_fw_update.yaml [hunk 5]
```diff
@@ -40,6 +40,21 @@ parameters:
     type: boolean
     default: False
     description: 'Priority tag required'
+  ESWITCH_HAIRPIN_TOT_BUFFER_SIZE:
+    type: json
+    default: {"*": "17"}
+    description: If "*" is on of the json keys, then its value will set to all indexes
```
Comment (adrianchiris) [inline templates/firstboot/mellanox_fw_update.yaml]: i would reword
`If a single key of "*" is provided, then its value will be set to all indexes`

File: templates/firstboot/mellanox_fw_update.yaml [hunk 6]
```diff
@@ -40,6 +40,21 @@ parameters:
...
+                 If you need to set configuration for specific indexes you can pass the
+                 value as below for index 2 to be 17 and index 3 to be 16
+                 Example, {"2": "17", "3": "16"}
+                 Make sure to choose only the PCI ends with 0
+  ESWITCH_HAIRPIN_DESCRIPTORS:
+    type: json
+    default: {"*": "11"}
+    description: If "*" is on of the json keys, then its value will set to all indexes
```
Comment (adrianchiris) [inline templates/firstboot/mellanox_fw_update.yaml]: i would reword
`If a single key of "*" is provided, then its value will be set to all indexes`

File: templates/firstboot/mellanox_fw_update.yaml [hunk 7]
```diff
@@ -193,16 +212,30 @@ resources:
...
+                                    d[0], first_index, last_index))
+                            for key, value in array_param_dict.items():
+                                r[key] = value
+                                continue
                         r[d[0]] = d[1]
                     return r
 
-                def get_device_conf_dict(self):
-                    """ Get device Configurations
+                def get_device_conf_dict(self, param_name=None):
```
Comment (adrianchiris) [inline templates/firstboot/mellanox_fw_update.yaml:236]: The flow is a bit convoluted, having `get_device_conf_dict` call `_mstconfig_parse_data` which calls back `get_device_conf_dict`

you could modify get_device_conf_dict to perform a 2 passess on the output. 
1 the normal mlxconfig query
2. expand the array configuration to their elements (you can do that recursively if you prefer using `get_device_conf_dict`)

File: templates/firstboot/mellanox_fw_update.yaml [hunk 8]
```diff
@@ -181,6 +199,7 @@ resources:
                 def __init__(self, pci_dev):
                     self.pci_dev = pci_dev
                     self._tool_confs = None
+                    self.arrayed_mlnx_config = {}
```
Comment (adrianchiris) [inline templates/firstboot/mellanox_fw_update.yaml]: maybe add a comment explaining what this attribute stores/represents ? or better yet just define a class for it (which has first and last index attributes)

also perhaps a better name would be:  `array_config_info`  ?

File: templates/firstboot/mellanox_fw_update.yaml [hunk 9]
```diff
@@ -193,16 +212,30 @@ resources:
                             break
                     for i in range(c, len(data)):
                         d = list(filter(None, data[i].strip().split()))
+                        array_value = re.match(ARRAY_VALUE_REGEX, d[1])
+                        if array_value:
+                            first_index = array_value.group('first_index')
+                            last_index =  array_value.group('last_index')
+                            self.arrayed_mlnx_config[d[0]] = [first_index, last_index]
```
Comment (adrianchiris) [inline templates/firstboot/mellanox_fw_update.yaml]: i prefer not to hide filling up the self.arrayed_mlnx_config in _mst_config_parse_data

can we have a separate method that gets the conf_dict and fills up that information ?

File: templates/firstboot/mellanox_fw_update.yaml [hunk 10]
```diff
@@ -230,7 +263,30 @@ resources:
                                       " please update to the latest mstflint package." % key)
                             continue
 
-                        if current_mlx_config.get(key) and value.lower(
+                        if key in self.arrayed_mlnx_config.keys():
+                            if value.get('*'):
```
Comment (adrianchiris) [inline templates/firstboot/mellanox_fw_update.yaml]: id prefer if you implement the logic to expand the user provided "conf_dict" in a separate method.
also  need to update the docstring for set_config.

Comment (adrianchiris) [reply templates/firstboot/mellanox_fw_update.yaml]: need to validate that if "*" is provided, no other entries are in the dict.

File: templates/firstboot/mellanox_fw_update.yaml [hunk 11]
```diff
@@ -230,7 +263,30 @@ resources:
...
+                                            param_value.lower() not in \
+                                            current_mlx_config.get(param_key).lower():
+                                        params_to_set.append("%s=%s" % (param_key,
+                                                                        param_value))
+                            else:
+                                 # Aggregate required configuration to specific indexes
+                                 for key_index, param_value in value.items():
+                                     param_key = "%s[%s]" % (key, key_index)
+                                     if current_mlx_config.get(param_key) and \
+                                             str(param_value).lower() not in \
```
Comment (adrianchiris) [inline templates/firstboot/mellanox_fw_update.yaml]: is there a reason you are using `in` and not `==` ?
e.g  "1" in "100" is true and would then lead to the configuration to be skipped

File: templates/firstboot/mellanox_fw_update.yaml [hunk 12]
```diff
@@ -230,7 +263,30 @@ resources:
...
+                                             current_mlx_config.get(param_key).lower():
+                                         params_to_set.append("%s=%s" % (param_key,
+                                                                         param_value))
+
+                        elif current_mlx_config.get(key) and value.lower(
                         ) not in current_mlx_config.get(key).lower():
```
Comment (adrianchiris) [inline templates/firstboot/mellanox_fw_update.yaml]: same issue is here as well i think

**Review by adrianchiris — CHANGES_REQUESTED** [review]

**Comment by adrianchiris** [comment]
@moshe010 @wmousa

File: templates/firstboot/mlnx_fw_update.py [hunk 13]
```diff
@@ -157,9 +180,77 @@ def param_supp_by_config_tool(self, param_name):
...
+            self._tool_confs = run_command(
+                "mstconfig", "-d", self.pci_dev, "i")
+        # trim any array index if present
+        indexed_param = re.match(ARRAY_PARAM_REGEX, param_name)
+        if indexed_param:
+            param_name = indexed_param.group('param_name')
         return param_name in self._tool_confs
 
+    def _build_config_param_metadata_map(self, conf_dict):
+        self.mlnx_config_array_param_metadata = dict()
```
Comment (moshe010) [inline templates/firstboot/mlnx_fw_update.py]: can you make all the code use {} instead of dict() just be consistent

**Review by moshe010 — COMMENTED** [review]
the change is very complex ;(. you probably tested only the script, right?
I guess Waleed will need to test  full deployment ...

File: templates/firstboot/mlnx_fw_update.py (hunk 13 above)
Comment (adrianchiris) [reply templates/firstboot/mlnx_fw_update.py]: sure

**Review by adrianchiris — COMMENTED** [review]
LGTM

**Review by moshe010 — APPROVED** [review]
